org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/DocumentSymbolHandler.java

```diff
@@ -147,7 +148,13 @@ private void collectChildren(ITypeRoot unit, IJavaElement[] elements, ArrayList<
 
 	private List<DocumentSymbol> getHierarchicalOutline(ITypeRoot unit, IProgressMonitor monitor) {
 		try {
-			return Stream.of(filter(unit.getChildren())).map(child -> toDocumentSymbol(child, monitor)).filter(Objects::nonNull).collect(Collectors.toList());
+			IJavaElement[] children = unit.getChildren();
+			Stream<IJavaElement> childrenStream = Stream.of(filter(children));
+			if (unit instanceof IClassFile) {
+				// Prepend Package element as the first child
+				childrenStream = Stream.concat(Stream.of(unit.getParent()), childrenStream);

```

Comment:
This change still won't fix class files but I think I see why.

Have a look at `unit.getParent()` for a class file. It is an `IPackageFragment` (eg. `JarPackageFragment` when the class file comes from a jar file). It's `getElementType()` value will not make it past the check at https://github.com/eclipse/eclipse.jdt.ls/pull/2619/files#diff-820919a8fb15f7287a49d35ff449241622cb4ac6bb83580d5b214b014bea2951R172 so that needs to be modified to accept a package fragment. Currently it would accept a package declaration, but that comes from the source file's model, and seems like it doesn't exist in the class file model.

Once that works you'll probably need to adjust `mapKind(..)` (https://github.com/eclipse/eclipse.jdt.ls/pull/2619/files#diff-820919a8fb15f7287a49d35ff449241622cb4ac6bb83580d5b214b014bea2951R266) to use the proper kind (which controls the icon used in the outline view).

Reply:
Working on the suggested changes.

Reply:
I have committed and pushed the changes.